File: 剑指offer/剑指 Offer 21. 调整数组顺序使奇数位于偶数前面/exchange .js
var exchange = function (nums) {
  let left = 0;
  let right = nums.length - 1;
  while (left < right) {
    if (nums[left] % 2 === 0 && nums[right] % 2 === 1) {
      [nums[left], nums[right]] = [nums[right], nums[left]];
      left++;
      right--;
    } else if (nums[left] % 2 === 0) {
      [nums[left], nums[right]] = [nums[right], nums[left]];
      right--;
    } else if (nums[right] % 2 === 1) {
      [nums[left], nums[right]] = [nums[right], nums[left]];
      left++;
    } else {
      left++;
      right--;
    }
  }
  return nums;
};

File: 剑指offer/剑指 Offer 21. 调整数组顺序使奇数位于偶数前面/exchange.py
# 解法一:
class Solution:
    def exchange(self, nums: List[int]) -> List[int]:
        slow = fast = 0
        while fast < len(nums):
            if nums[fast] & 1:
                nums[slow], nums[fast] = nums[fast], nums[slow]
                slow += 1
            fast += 1
        return nums


# 解法二:
class Solution:
    def exchange(self, nums: List[int]) -> List[int]:
        left, right = 0 ,len(nums) - 1
        while left < right:
            if nums[left] & 1:
                left+=1
                continue
            if nums[right] & 1 == 0:
                right-=1
                continue
            nums[left], nums[right] = nums[right], nums[left]
            
        return nums